File: src/api/endpoints/runners/lambdas/runner_starter.py
"""Runner starter Lambda handler - routes job requests to runner APIs."""

import json
import logging
import os
import time
import urllib.request
import urllib.error
from typing import Any

from botocore.exceptions import ClientError

from common.aws_clients import get_ssm_client
from common.cloudwatch import publish_metric
from common.lambda_utils import get_sqs_records, empty_records_response, parse_error_response
from runner_labels import get_runner_type_from_labels

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

METRICS_NAMESPACE = "RunnerStarter"

_cache: dict[str, Any] = {"api_key": None}

_circuit_breaker_state: dict[str, Any] = {
    "failures": 0,
    "last_failure_time": 0.0,
    "state": "closed",
}


def _check_circuit_breaker() -> bool:
    """Check if requests should be allowed through the circuit breaker.

    Returns:
        True if requests should proceed, False if circuit is open
    """
    current_time = time.time()
    failure_threshold = 5
    timeout_seconds = 60

    if _circuit_breaker_state["state"] == "open":
        if current_time - _circuit_breaker_state["last_failure_time"] > timeout_seconds:
            logger.info("Circuit breaker transitioning to half-open state")
            _circuit_breaker_state["state"] = "half-open"
            _circuit_breaker_state["failures"] = 0
            publish_metric(METRICS_NAMESPACE, "CircuitBreakerState", 1.0, "Count")
            return True
        publish_metric(METRICS_NAMESPACE, "CircuitBreakerState", 2.0, "Count")
        return False

    if _circuit_breaker_state["failures"] >= failure_threshold:
        logger.warning(
            "Circuit breaker opening due to %d failures",
            _circuit_breaker_state["failures"],
        )
        _circuit_breaker_state["state"] = "open"
        _circuit_breaker_state["last_failure_time"] = current_time
        publish_metric(METRICS_NAMESPACE, "CircuitBreakerState", 2.0, "Count")
        return False

    publish_metric(METRICS_NAMESPACE, "CircuitBreakerState", 0.0, "Count")
    return True


def _record_circuit_breaker_success() -> None:
    """Record a successful request for the circuit breaker."""
    if _circuit_breaker_state["state"] == "half-open":
        logger.info("Circuit breaker closing after successful request")
        _circuit_breaker_state["state"] = "closed"
    _circuit_breaker_state["failures"] = 0


def _record_circuit_breaker_failure() -> None:
    """Record a failed request for the circuit breaker."""
    _circuit_breaker_state["failures"] += 1
    _circuit_breaker_state["last_failure_time"] = time.time()
    if _circuit_breaker_state["state"] == "half-open":
        logger.warning(
            "Circuit breaker reopening after failed request in half-open state"
        )
        _circuit_breaker_state["state"] = "open"


def _should_record_circuit_breaker_failure(status_code: int | None) -> bool:
    """Check if a status code should count as a circuit breaker failure."""
    return status_code is None


def _get_api_key(force_refresh: bool = False) -> str:
    """Get API key from SSM Parameter Store with caching.

    Args:
        force_refresh: Force refresh of cached key

    Returns:
        API key string

    Raises:
        RuntimeError: If API key cannot be retrieved
    """
    if force_refresh:
        _cache["api_key"] = None

    if _cache["api_key"]:
        return _cache["api_key"]

    parameter_name = os.environ.get("API_KEY_PARAMETER_NAME")
    if not parameter_name:
        raise RuntimeError("API_KEY_PARAMETER_NAME environment variable not set")

    try:
        response = get_ssm_client().get_parameter(
            Name=parameter_name, WithDecryption=True
        )
        api_key = response.get("Parameter", {}).get("Value", "")
        _cache["api_key"] = api_key
        return api_key
    except ClientError as err:
        logger.error("Failed to retrieve API key: %s", str(err))
        raise RuntimeError(f"Cannot retrieve API key: {err}") from err


def _make_http_request_with_retry(
    endpoint: str,
    payload: dict[str, Any],
    headers: dict[str, str] | None = None,
    max_retries: int = 3,
) -> tuple[bool, dict[str, Any] | None, str | None, int | None]:
    """Make HTTP POST request with exponential backoff retry.

    Args:
        endpoint: Full URL to POST to
        payload: JSON payload
        headers: Additional headers
        max_retries: Maximum retry attempts

    Returns:
        Tuple of (success, response_data, error, status_code)
    """
    request_headers = {"Content-Type": "application/json"}
    if headers:
        request_headers.update(headers)

    last_status_code: int | None = None

    for attempt in range(max_retries + 1):
        try:
            request = urllib.request.Request(
                endpoint,
                data=json.dumps(payload).encode("utf-8"),
                headers=request_headers,
                method="POST",
            )

            with urllib.request.urlopen(request, timeout=30) as response:
                last_status_code = response.status
                if 200 <= response.status < 300:
                    response_body = response.read().decode("utf-8")
                    response_data = json.loads(response_body) if response_body else {}
                    return True, response_data, None, response.status
        except urllib.error.HTTPError as err:
            last_status_code = err.code
            if 400 <= err.code < 500:
                return False, None, f"HTTP {err.code}", err.code
            if attempt >= max_retries:
                return (
                    False,
                    None,
                    f"HTTP {err.code} after {max_retries + 1} attempts",
                    err.code,
                )
        except urllib.error.URLError as err:
            if attempt >= max_retries:
                return (
                    False,
                    None,
                    f"{err.reason} after {max_retries + 1} attempts",
                    last_status_code,
                )

        # Exponential backoff
        time.sleep(2**attempt)

    return False, None, "Max retries exceeded", last_status_code


def _build_runner_endpoint(endpoint_suffix: str) -> str:
    """Build full runner endpoint URL."""
    base_url = os.environ.get("API_BASE_URL", "")
    return f"{base_url}/v1/{endpoint_suffix}"


def _handle_route_success(
    job_id: int, runner_type: str, response_data: dict[str, Any] | None
) -> dict[str, Any]:
    """Handle successful route request."""
    logger.info("Successfully routed job %s to %s runner", job_id, runner_type)
    _record_circuit_breaker_success()
    return {"success": True, "runner_type": runner_type, "response": response_data}


def _handle_route_failure(
    job_id: int, error: str | None, status_code: int | None
) -> dict[str, Any]:
    """Handle failed route request."""
    logger.error("Failed to route job %s: %s", job_id, error)
    if _should_record_circuit_breaker_failure(status_code):
        _record_circuit_breaker_failure()
    else:
        logger.warning(
            "Status %s for job %s - not counting as circuit breaker failure",
            status_code,
            job_id,
        )
    return {"success": False, "error": error}


def _route_runner_request(
    job_id: int,
    job_labels: list[str],
    github_repo: str,
    run_id: int | None = None,
) -> dict[str, Any]:
    """Route a job request to the appropriate runner API.

    Args:
        job_id: GitHub job ID
        job_labels: Job labels
        github_repo: Repository full name
        run_id: Optional workflow run ID

    Returns:
        Result dictionary with success status
    """
    if not _check_circuit_breaker():
        logger.error("Circuit breaker is open, rejecting request for job %s", job_id)
        return {
            "success": False,
            "error": "Service temporarily unavailable (circuit breaker open)",
        }

    runner_type, endpoint_suffix = get_runner_type_from_labels(job_labels)
    if not runner_type or not endpoint_suffix:
        logger.error("No matching runner type for labels: %s", job_labels)
        return {
            "success": False,
            "error": f"No matching runner type for labels: {json.dumps(job_labels)}",
        }

    try:
        api_key = _get_api_key()
    except RuntimeError as err:
        logger.error("Cannot route job %s: %s", job_id, str(err))
        return {"success": False, "error": str(err)}

    endpoint = _build_runner_endpoint(endpoint_suffix)
    payload = {
        "job_id": job_id,
        "job_labels": job_labels,
        "github_repo": github_repo,
        "run_id": run_id,
        "runner_type": runner_type,
    }

    logger.info(
        "Routing job %s to %s runner: %s (run_id=%s)",
        job_id,
        runner_type,
        endpoint,
        run_id,
    )

    success, response_data, error, status_code = _make_http_request_with_retry(
        endpoint, payload, {"x-api-key": api_key}
    )

    if success:
        return _handle_route_success(job_id, runner_type, response_data)
    return _handle_route_failure(job_id, error, status_code)


def _handle_sqs_message(message: dict[str, Any]) -> dict[str, Any]:
    """Handle a single SQS message.

    Args:
        message: SQS message record

    Returns:
        Result dictionary with success status
    """
    try:
        body = json.loads(message.get("body", "{}"))
        job_id = body.get("job_id")
        job_labels = body.get("job_labels", [])
        github_repo = body.get("github_repo")
        run_id = body.get("run_id")

        logger.info(
            "Processing job from SQS: job_id=%s, labels=%s, repo=%s, run_id=%s",
            job_id,
            json.dumps(job_labels),
            github_repo,
            run_id,
        )

        result = _route_runner_request(job_id, job_labels, github_repo, run_id)

        if result.get("success"):
            logger.info("Successfully processed SQS message for job %s", job_id)
            return {"success": True}

        logger.error(
            "Failed to process SQS message for job %s: %s",
            job_id,
            result.get("error"),
        )
        return {"success": False, "error": result.get("error")}
    except (json.JSONDecodeError, KeyError) as err:
        return parse_error_response(err)


def lambda_handler(event: dict[str, Any], _context: Any) -> dict[str, Any]:
    """Process queued workflow jobs to start runners.

    Receives SQS messages containing GitHub workflow job events and provisions
    the appropriate runner infrastructure (ECS task or EC2 instance).
    """
    handler_start = time.time()
    sqs_records = get_sqs_records(event)
    if not sqs_records:
        return empty_records_response()

    logger.info("Processing %d job(s) from SQS", len(sqs_records))
    results = [_handle_sqs_message(record) for record in sqs_records]

    elapsed_ms = (time.time() - handler_start) * 1000
    publish_metric(METRICS_NAMESPACE, "ProcessingTime", elapsed_ms, "Milliseconds")

    if not all(r.get("success") for r in results):
        raise RuntimeError("One or more job messages failed")

    return {
        "statusCode": 200,
        "body": json.dumps({"message": "Processed", "count": len(sqs_records)}),
    }

File: src/api/endpoints/runners/lambdas/runner_starter.js
const https = require('https');
const { getSSMClient } = require('/opt/nodejs/clients/ssm');
const { getCloudWatchClient } = require('/opt/nodejs/clients/cloudwatch');
const {
  parseLabels,
  validateLabels,
  LabelParseError,
  LabelValidationError
} = require('/opt/nodejs/common');
const { GetParameterCommand } = require('@aws-sdk/client-ssm');
const { PutMetricDataCommand } = require('@aws-sdk/client-cloudwatch');

const logger = {
  info: (...args) => console.log('[INFO]', ...args),
  warn: (...args) => console.warn('[WARN]', ...args),
  error: (...args) => console.error('[ERROR]', ...args)
};

const apiKeyCache = { value: null };

const circuitBreakerState = {
  failures: 0,
  lastFailureTime: 0,
  state: 'closed'
};

async function publishMetric(metricName, value, unit = 'None') {
  try {
    await getCloudWatchClient().send(new PutMetricDataCommand({
      Namespace: 'RunnerStarter',
      MetricData: [{
        MetricName: metricName,
        Value: value,
        Unit: unit,
        Timestamp: new Date()
      }]
    }));
  } catch (err) {
    logger.warn('Failed to publish metric', metricName, ':', err.message);
  }
}

function checkCircuitBreaker() {
  const currentTime = Date.now() / 1000;
  const failureThreshold = 5;
  const timeoutSeconds = 60;

  if (circuitBreakerState.state === 'open') {
    if (currentTime - circuitBreakerState.lastFailureTime > timeoutSeconds) {
      logger.info('Circuit breaker transitioning to half-open state');
      circuitBreakerState.state = 'half-open';
      circuitBreakerState.failures = 0;
      publishMetric('CircuitBreakerState', 1.0, 'Count');
      return true;
    }
    publishMetric('CircuitBreakerState', 2.0, 'Count');
    return false;
  }

  if (circuitBreakerState.failures >= failureThreshold) {
    logger.warn('Circuit breaker opening due to', circuitBreakerState.failures, 'failures');
    circuitBreakerState.state = 'open';
    circuitBreakerState.lastFailureTime = currentTime;
    publishMetric('CircuitBreakerState', 2.0, 'Count');
    return false;
  }

  publishMetric('CircuitBreakerState', 0.0, 'Count');
  return true;
}

function recordCircuitBreakerSuccess() {
  if (circuitBreakerState.state === 'half-open') {
    logger.info('Circuit breaker closing after successful request');
    circuitBreakerState.state = 'closed';
  }
  circuitBreakerState.failures = 0;
}

function recordCircuitBreakerFailure() {
  circuitBreakerState.failures++;
  circuitBreakerState.lastFailureTime = Date.now() / 1000;
  if (circuitBreakerState.state === 'half-open') {
    logger.warn('Circuit breaker reopening after failed request in half-open state');
    circuitBreakerState.state = 'open';
  }
}

function shouldRecordCircuitBreakerFailure(statusCode) {
  return statusCode === null;
}

async function getApiKey(forceRefresh = false) {
  if (forceRefresh) {
    apiKeyCache.value = null;
  }
  if (apiKeyCache.value) {
    return apiKeyCache.value;
  }

  const parameterName = process.env.API_KEY_PARAMETER_NAME;
  if (!parameterName) {
    throw new Error('API_KEY_PARAMETER_NAME environment variable not set');
  }

  try {
    const response = await getSSMClient().send(new GetParameterCommand({
      Name: parameterName,
      WithDecryption: true
    }));
    const apiKey = response.Parameter?.Value;
    apiKeyCache.value = apiKey;
    return apiKey;
  } catch (err) {
    logger.error('Failed to retrieve API key:', err.message);
    throw new Error(`Cannot retrieve API key: ${err.message}`);
  }
}

function makeHttpRequestWithRetry(endpoint, payload, headers = {}, maxRetries = 3) {
  return new Promise((resolve) => {
    const url = new URL(endpoint);
    let lastStatusCode = null;

    const attempt = (attemptNum) => {
      const options = {
        hostname: url.hostname,
        port: url.port || 443,
        path: url.pathname + url.search,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...headers
        }
      };

      const req = https.request(options, (res) => {
        let data = '';
        res.on('data', chunk => { data += chunk; });
        res.on('end', () => {
          lastStatusCode = res.statusCode;
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve([true, JSON.parse(data || '{}'), null, res.statusCode]);
          } else if (res.statusCode >= 400 && res.statusCode < 500) {
            resolve([false, null, `HTTP ${res.statusCode}`, res.statusCode]);
          } else if (attemptNum >= maxRetries) {
            resolve([false, null, `HTTP ${res.statusCode} after ${maxRetries + 1} attempts`, res.statusCode]);
          } else {
            setTimeout(() => attempt(attemptNum + 1), Math.pow(2, attemptNum) * 1000);
          }
        });
      });

      req.on('error', (err) => {
        if (attemptNum >= maxRetries) {
          resolve([false, null, `${err.message} after ${maxRetries + 1} attempts`, lastStatusCode]);
        } else {
          setTimeout(() => attempt(attemptNum + 1), Math.pow(2, attemptNum) * 1000);
        }
      });

      req.write(JSON.stringify(payload));
      req.end();
    };

    attempt(0);
  });
}

function getRunnerTypeFromLabels(jobLabels) {
  let runnerType = null;
  let endpointSuffix = null;

  try {
    const parsed = parseLabels(jobLabels);
    validateLabels(parsed);
    const isE2e = jobLabels.includes('e2e');

    if (parsed.platform === 'ec2') {
      runnerType = isE2e ? 'ec2-e2e' : 'ec2';
      endpointSuffix = 'ec2-runner';
    } else if (parsed.platform === 'ecs') {
      runnerType = isE2e ? 'fargate-e2e' : 'fargate';
      endpointSuffix = 'ecs-runner';
    }
  } catch (err) {
    if (!(err instanceof LabelParseError || err instanceof LabelValidationError)) {
      throw err;
    }
  }

  return [runnerType, endpointSuffix];
}

function buildRunnerEndpoint(endpointSuffix) {
  return `${process.env.API_BASE_URL}/v1/${endpointSuffix}`;
}

function handleRouteSuccess(jobId, runnerType, responseData) {
  logger.info('Successfully routed job', jobId, 'to', runnerType, 'runner');
  recordCircuitBreakerSuccess();
  return { success: true, runner_type: runnerType, response: responseData };
}

function handleRouteFailure(jobId, error, statusCode) {
  logger.error('Failed to route job', jobId, ':', error);
  if (shouldRecordCircuitBreakerFailure(statusCode)) {
    recordCircuitBreakerFailure();
  } else {
    logger.warn('Status', statusCode, 'for job', jobId, '- not counting as circuit breaker failure');
  }
  return { success: false, error };
}

async function routeRunnerRequest(jobId, jobLabels, githubRepo, runId = null) {
  if (!checkCircuitBreaker()) {
    logger.error('Circuit breaker is open, rejecting request for job', jobId);
    return { success: false, error: 'Service temporarily unavailable (circuit breaker open)' };
  }

  const [runnerType, endpointSuffix] = getRunnerTypeFromLabels(jobLabels);
  if (!runnerType || !endpointSuffix) {
    logger.error('No matching runner type for labels:', jobLabels);
    return { success: false, error: `No matching runner type for labels: ${JSON.stringify(jobLabels)}` };
  }

  let apiKey;
  try {
    apiKey = await getApiKey();
  } catch (err) {
    logger.error('Cannot route job', jobId, ':', err.message);
    return { success: false, error: err.message };
  }

  const endpoint = buildRunnerEndpoint(endpointSuffix);
  const payload = {
    job_id: jobId,
    job_labels: jobLabels,
    github_repo: githubRepo,
    run_id: runId,
    runner_type: runnerType
  };

  logger.info('Routing job', jobId, 'to', runnerType, 'runner:', endpoint, '(run_id=', runId, ')');

  const [success, responseData, error, statusCode] = await makeHttpRequestWithRetry(
    endpoint, payload, { 'x-api-key': apiKey }
  );

  if (success) {
    return handleRouteSuccess(jobId, runnerType, responseData);
  }
  return handleRouteFailure(jobId, error, statusCode);
}

async function handleSqsMessage(message) {
  try {
    const body = JSON.parse(message.body);
    const { job_id: jobId, job_labels: jobLabels, github_repo: githubRepo, run_id: runId } = body;

    logger.info(
      'Processing job from SQS: job_id=', jobId, ', labels=', JSON.stringify(jobLabels),
      ', repo=', githubRepo, ', run_id=', runId
    );

    const result = await routeRunnerRequest(jobId, jobLabels, githubRepo, runId);

    if (result.success) {
      logger.info('Successfully processed SQS message for job', jobId);
      return { success: true };
    }

    logger.error('Failed to process SQS message for job', jobId, ':', result.error);
    return { success: false, error: result.error };
  } catch (err) {
    logger.error('Failed to parse SQS message:', err.message);
    return { success: false, error: `Invalid message format: ${err.message}` };
  }
}

exports.handler = async (event, _context) => {
  const startTime = Date.now();
  logger.info('Received event:', JSON.stringify(event));

  const records = event.Records || [];
  if (records.length === 0) {
    logger.warn('No records in event');
    return { statusCode: 200, body: JSON.stringify({ message: 'No records to process' }) };
  }

  logger.info('Processing', records.length, 'job(s) from SQS');

  const results = [];
  for (const record of records) {
    const result = await handleSqsMessage(record);
    results.push(result);
  }

  const elapsed = Date.now() - startTime;
  await publishMetric('ProcessingTime', elapsed, 'Milliseconds');

  if (!results.every(r => r.success)) {
    throw new Error('One or more job messages failed');
  }

  return { statusCode: 200, body: JSON.stringify({ message: 'Processed', count: records.length }) };
};

// Exported for testing
module.exports = {
  handler: exports.handler,
  getApiKey,
  checkCircuitBreaker,
  recordCircuitBreakerSuccess,
  recordCircuitBreakerFailure,
  routeRunnerRequest,
  handleSqsMessage,
  getRunnerTypeFromLabels,
  makeHttpRequestWithRetry
};
